inference/inference-core/src/commonMain/kotlin/io/kinference.core/operators/convolution/Conv.kt

```diff
@@ -0,0 +1,278 @@
+                    var pad = (outputShape - 1) * strides[i % size] + wShape[i % size + 2] - xShape[i % size + 2]
+                    if (i < size && pad % 2 == 1) // not necessary to check whether pad is odd or not
+                        pad++
+                    pad / 2
+                }
+            }
+
+            "VALID" -> {
+                xShape.forEachIndexed { index, i -> require(i >= wShape[index]) }
+
+                IntArray(xShape.size * 2) { _ -> 0 }
+            }
+
+            "NOTSET" -> {
+                if (pads == null)
+                    IntArray(xShape.size * 2) { _ -> 0 }
+                else
+                    pads as IntArray
+            }
+
+            else -> throw IllegalArgumentException("Invalid auto_pad argument: $autoPad.")
+        }
+    }
+}
+
+/***
+ * Extends shape of tensor with given pads. Ignores 1st and 2nd dimension.
+ */
+private fun getShapeWithPads(shape: IntArray, pads: IntArray) =
+    shape.mapIndexed { index, i -> if (index < 2) i else i + pads[index - 2] + pads[index + shape.size - 4] }.toIntArray()
+
+/***
+ * Extends shape of tensor with given dilations. Ignores 1st and 2nd dimension.
+ */
+private fun getShapeWithDilations(shape: IntArray, dilations: IntArray) =
+    shape.mapIndexed { index, i -> if (index < 2) i else (i - 1) * dilations[index - 2] + 1 }.toIntArray()
+
+
+private fun FloatNDArray.convolve(
+    w: FloatNDArray,
+    result: MutableFloatNDArray,
+    pads: IntArray, strides:
+    IntArray,
+    dilations: IntArray,
+    groups: Int,
+    item: Int,
+    feature: Int
+) {
+    val wModShape = getShapeWithDilations(w.shape, dilations)
+    val xShape = this.shape - wModShape + IntArray(this.shape.size) { 1 }
+    xShape[1] = this.shape[1]
+
+    for (i in 0 until this.shape[1]) {
+        val xIterator = TensorIterator(xShape, item, i, pads, strides)
+
+        val outFeature = feature + (i / w.shape[1]) * (w.shape[0] / groups)
+        val resultIterator = TensorIterator(result.shape, item, outFeature)
+        while (xIterator.hasNext()) {
+            val wIterator = TensorIterator(w.shape, outFeature, i % w.shape[1])
+
+            val xPos = xIterator.next()
+            val resultPos = resultIterator.next()
+
+            while (wIterator.hasNext()) {
+                val wPos = wIterator.next()
+                val weight = w[wPos]
+
+                val shift = wPos.mapIndexed { index, value -> if (index < 2) 0 else value * dilations[index - 2] }.toIntArray()
+
+                result[resultPos] += this.getWithPads(xPos + shift) * weight
+            }
+        }
+    }
+}
+
+private fun FloatNDArray.getWithPads(indexes: IntArray): Float {
+    if (indexes.any { i -> i < 0 })
+        return 0f
+    indexes.forEachIndexed { index, i -> if (i >= this.shape[index]) return 0f }
+    return this[indexes]
+}
+
+private operator fun IntArray.plus(oth: IntArray): IntArray {
+    return this.mapIndexed { index, i -> i + oth[index] }.toIntArray()
```

Comment:
Better to avoid excessive memory allocations as map function create List. If you can easily rewrite the function with simple array creation, do that instead. IntArray(this.size) { this[it] + oth[it] }